rust/src/message/traits.rs

```diff
@@ -0,0 +1,400 @@
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> LengthValueBuffer<&'a T> {
+    /// Get a reference to the value field.
+    ///
+    /// # Panic
+    ///
+    /// This method may panic if buffer is not a valid Length-Value item.
+    pub fn value(&self) -> &'a [u8] {
+        &self.inner.as_ref()[self.value_range()]
+    }
+
+    /// Get a reference to the underlying buffer.
+    ///
+    /// # Panic
+    ///
+    /// This method may panic if buffer is not a valid Length-Value item.
+    pub fn bytes(self) -> &'a [u8] {
+        let range = self.value_range();
+        &self.inner.as_ref()[..range.end]
+    }
+}
+
+macro_rules! impl_traits_for_length_value_types {
+    ($ty: ty) => {
+        impl ToBytes for $ty {
+            fn buffer_length(&self) -> usize {
+                LENGTH_FIELD.end + self.as_ref().len()
+            }
+
+            fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+                let mut writer = LengthValueBuffer::new_unchecked(buffer.as_mut());
+                writer.set_length(self.buffer_length() as u32);
+                writer.value_mut().copy_from_slice(self.as_ref());
+            }
+        }
+
+        impl FromBytes for $ty {
+            fn from_bytes<T: AsRef<[u8]>>(buffer: &T) -> Result<Self, DecodeError> {
+                let reader = LengthValueBuffer::new(buffer.as_ref())?;
+                Ok(Self::from(reader.value()))
+            }
+        }
+    };
+}
+
+impl_traits_for_length_value_types!(crate::certificate::Certificate);
+impl_traits_for_length_value_types!(crate::mask::Mask);
+impl_traits_for_length_value_types!(crate::mask::MaskedModel);
+
+const ENTRY_LENGTH: usize = SumParticipantPublicKey::LENGTH + EncrMaskSeed::BYTES;
+
+impl ToBytes for LocalSeedDict {
+    fn buffer_length(&self) -> usize {
+        LENGTH_FIELD.end + self.len() * ENTRY_LENGTH
+    }
+
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+        let mut writer = Cursor::new(buffer.as_mut());
+        let length = self.buffer_length() as u32;
+        writer.write(&length.to_be_bytes()).unwrap();
+        for (key, value) in self {
+            writer.write(key.as_slice()).unwrap();
+            writer.write(value.as_ref()).unwrap();
+        }
+    }
+}
+
+impl FromBytes for LocalSeedDict {
+    fn from_bytes<T: AsRef<[u8]>>(buffer: &T) -> Result<Self, DecodeError> {
+        let reader = LengthValueBuffer::new(buffer.as_ref())?;
+        let nb_entries = reader.value_length() as usize / ENTRY_LENGTH;
+        let mut dict = LocalSeedDict::with_capacity(nb_entries);
+
+        let key_length = SumParticipantPublicKey::LENGTH;
+        let mut entries = reader.value().chunks_exact(ENTRY_LENGTH);
+        for chunk in &mut entries {
+            // safe unwraps: lengths of slices are guaranteed
+            // by constants.
+            let key = SumParticipantPublicKey::from_slice(&chunk[..key_length]).unwrap();
+            let value = EncrMaskSeed::try_from(&chunk[key_length..]).unwrap();
```

Comment:
i double-checked with my mr to avoid merge conflicts, basically you can ignore my changes to the deserialization of the local seed dict, it's all handled here as well.

in addition to that, the second line will change once my mr is merged to
```
let value = EncryptedMaskSeed::from_slice(&chunk[key_length..]).unwrap();
```
or better by using the `ByteObject` both lines can be changed to
```
let key = SumParticipantPublicKey::from_slice_unchecked(&chunk[..key_length]);
let value = EncryptedMaskSeed::from_slice_unchecked(&chunk[key_length..]);
```